src/cuda/backend.f90

```diff
@@ -0,0 +1,423 @@
+      call self%transeq_cuda_dist(dv, du, dw, v, u, w, dirps, &
+                                  self%yblocks, self%ythreads)
+
+   end subroutine transeq_y_cuda
+
+   subroutine transeq_z_cuda(self, du, dv, dw, u, v, w, dirps)
+      implicit none
+
+      class(cuda_backend_t) :: self
+      class(field_t), intent(inout) :: du, dv, dw
+      class(field_t), intent(in) :: u, v, w
+      type(dirps_t), intent(in) :: dirps
+
+      ! u, v, w is reordered so that we pass w, u, v
+      call self%transeq_cuda_dist(dw, du, dv, w, u, v, dirps, &
+                                  self%zblocks, self%zthreads)
+
+   end subroutine transeq_z_cuda
+
+   subroutine transeq_cuda_dist(self, du, dv, dw, u, v, w, dirps, &
+                                blocks, threads)
+      implicit none
+
+      class(cuda_backend_t) :: self
+      class(field_t), intent(inout) :: du, dv, dw
+      class(field_t), intent(in) :: u, v, w
+      type(dirps_t), intent(in) :: dirps
+      type(dim3), intent(in) :: blocks, threads
+
+      class(field_t), pointer :: temp_du, temp_duu, temp_d2u, &
+                                 temp_dv, temp_dvu, temp_d2v, &
+                                 temp_dw, temp_dwu, temp_d2w
+
+      real(dp), device, pointer, dimension(:, :, :) :: &
+         du_dev, duu_dev, d2u_dev, &
+         dv_dev, dvu_dev, d2v_dev, &
+         dw_dev, dwu_dev, d2w_dev
+
+      real(dp), device, pointer, dimension(:, :, :) :: u_dev, v_dev, w_dev, &
+                                                       ru_dev, rv_dev, rw_dev
+
+      type(cuda_tdsops_t), pointer :: der1st, der1st_sym, der2nd, der2nd_sym
+
+      select type(u); type is (cuda_field_t); u_dev => u%data_d; end select
+      select type(v); type is (cuda_field_t); v_dev => v%data_d; end select
+      select type(w); type is (cuda_field_t); w_dev => w%data_d; end select
+
+      select type(du); type is (cuda_field_t); ru_dev => du%data_d; end select
+      select type(dv); type is (cuda_field_t); rv_dev => dv%data_d; end select
+      select type(dw); type is (cuda_field_t); rw_dev => dw%data_d; end select
+
+      select type (tdsops => dirps%der1st)
+      type is (cuda_tdsops_t); der1st => tdsops
+      end select
+      select type (tdsops => dirps%der1st_sym)
+      type is (cuda_tdsops_t); der1st_sym => tdsops
+      end select
+      select type (tdsops => dirps%der2nd)
+      type is (cuda_tdsops_t); der2nd => tdsops
+      end select
+      select type (tdsops => dirps%der2nd_sym)
+      type is (cuda_tdsops_t); der2nd_sym => tdsops
+      end select
+
+      ! Copy halo data into buffer arrays
+      !self%u_send_s_dev(:, :, :) = u_dev(:, 1:4, :)
+      !self%u_send_e_dev(:, :, :) = u_dev(:, dirps%n - 3:dirps%n, :)
+      !self%v_send_s_dev(:, :, :) = v_dev(:, 1:4, :)
+      !self%v_send_e_dev(:, :, :) = v_dev(:, dirps%n - 3:dirps%n, :)
+      !self%w_send_s_dev(:, :, :) = w_dev(:, 1:4, :)
+      !self%w_send_e_dev(:, :, :) = w_dev(:, dirps%n - 3:dirps%n, :)
```

Comment:
To delete? or explain why it is commented out

Reply:
We need these assignments for the halo data exchange,  these arrays are the buffer arrays for sending halo data. NVIDIA compiler complained for this assignment, I guess because we're using the members of the class via `self%`. I need to move this into a function and then if I pass the buffer arrays as device arrays the assignment in the function should work fine just like in the tests.